runtime/lib/backend/LightningKokkosSimulator.cpp

```diff
@@ -131,58 +131,105 @@ void LightningKokkosSimulator::NamedOperation(const std::string &name,
+    Kokkos::deep_copy(gate_matrix, UnmanagedComplexHostView(matrix_kok.data(), matrix_kok.size()));
+
+    // Update the state-vector
+    this->device_sv->applyMultiQubitOp(gate_matrix, dev_wires, inverse);
+
+    // Update tape caching if required
+    if (this->cache_recording) {
+        this->cache_manager.addOperation("MatrixOp", {}, dev_wires, inverse);
+    }
 }
 
 auto LightningKokkosSimulator::Observable(ObsId id, const std::vector<std::complex<double>> &matrix,
                                           const std::vector<QubitIdType> &wires) -> ObsIdType
 {
-    if (id == ObsId::Hermitian) {
-        throw std::logic_error(
-            "Hermitian observable not implemented in PennyLane-Lightning-Kokkos");
-    }
-
-    QFailIf(wires.size() > GetNumQubits(), "Invalid number of wires");
+    QFailIf(wires.size() > this->GetNumQubits(), "Invalid number of wires");
     QFailIf(!isValidQubits(wires), "Invalid given wires");
 
     auto &&dev_wires = getDeviceWires(wires);
+
+    if (id == ObsId::Hermitian) {
+        return this->obs_manager.createHermitianObs(matrix, dev_wires);
+    }
+
     return this->obs_manager.createNamedObs(id, dev_wires);
 }
 
 auto LightningKokkosSimulator::TensorObservable(const std::vector<ObsIdType> &obs) -> ObsIdType
 {
-    throw std::logic_error("Tensor observable not implemented in PennyLane-Lightning-Kokkos");
+    return this->obs_manager.createTensorProdObs(obs);
 }
 
 auto LightningKokkosSimulator::HamiltonianObservable(const std::vector<double> &coeffs,
                                                      const std::vector<ObsIdType> &obs) -> ObsIdType
 {
-    throw std::logic_error("Hamiltonian observable not implemented in PennyLane-Lightning-Kokkos");
+    return this->obs_manager.createHamiltonianObs(coeffs, obs);
 }
 
 auto LightningKokkosSimulator::Expval(ObsIdType obsKey) -> double
 {
+    using UnmanagedComplexHostView = Kokkos::View<Kokkos::complex<double> *, Kokkos::HostSpace,
+                                                  Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
     QFailIf(!this->obs_manager.isValidObservables({obsKey}), "Invalid key for cached observables");
-    auto &&[obs, wires] = this->obs_manager.getObservable(obsKey);
 
     // update tape caching
     if (this->cache_recording) {
         cache_manager.addObservable(obsKey, Lightning::Measurements::Expval);
     }
 
-    if (obs == ObsId::PauliX) {
-        return this->device_sv->getExpectationValuePauliX(wires);
-    }
-    else if (obs == ObsId::PauliY) {
-        return this->device_sv->getExpectationValuePauliY(wires);
+    auto &&[obs, obs_type] = this->obs_manager.getObservable(obsKey);
+
+    if (obs_type != ObsType::Basic) {
+        std::cerr << "Warning: Expectation value of none-basic observables not supported yet in "
+                     "PennyLane-Lightning-Kokkos; the given observable is recorded for computing "
+                     "the runtime gradient algorithm."
+                  << std::endl;
```

Comment:
Yes! It's so tricky :/ Lightning-Kokkos has support for computing the gradient of circuits with `expval(Hamiltonian)` and `expval(TensorProd)` but doesn't provide a direct support for the computation of expectation values of these two observables. Their support is basically throughout passing the matrix of such observables using `qml.matrix`. I'll spend some time today checking if I can get a workaround this until they provide the C++ support in the next version.

Here, I didn't return an error to get the record-manager caches the `expval(Hamiltonian)` and `expval(TensorProd)` for gradient.

Reply:
wait, this is weird no? to support `(d/dx) expval(H(x))`, you should need to implement `expval(H(x))` as a first step?

Tagging @mlxd for context here

Reply:
Pretty sure we support this https://github.com/PennyLaneAI/pennylane-lightning-kokkos/blob/44aff20471ada590cd833d98b56553dc1e5799fc/pennylane_lightning_kokkos/lightning_kokkos.py#LL535C16-L535C16

It may be that @maliasadi needs to make use of the calling convention used to access the C++ layer.

Reply:
Yes I was surprised as well, but apparently this is the case 😄

Reply:
Bare in mind too that none of the devices were built to support external API use, so we may need to apln cross-squad efforts here to allow this in a dependable way.

Reply:
Looks like what you linked to @mlxd is a Python workaround for a Hamiltonian observable. However, Catalyst cannot use this method since in general the values needed to compute the matrix of such an observable are not available at compile time, and we don't have an implementation of `qml.matrix` that we can invoke to compute it at runtime.

Reply:
Yea, I mean that we bind to the C++ calls through Python. The bound names should be matched through the methods there.

Reply:
The 3 function below https://github.com/PennyLaneAI/pennylane-lightning-kokkos/blob/main/pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp#L1072 are the ones we call through the bindings. If you need anything different than these, additional mehtods may be needed. If you have a sparse data rep, it'd be the most performant. If not, passing in the numeric H values would be the next call. If you need H decompositions, that'll be a new method.

Reply:
No! This is not going to work in our end :upside_down_face: 

There is indeed a missing functionality in Lighnting-Kokkos to calculate the expval of generic observables. I've added a Kokkos kernel to Lighnting-Kokkos providing this missing support and now we can calculate the expval of every observables including Hamiltonian and TensorProd.

Reply:
I've added this patch as an extension to the current PR. 

I will create a PR in Lightning-Kokkos adding this new support and will adjust the changes in the runtime when that PR is successfully merged and included in the next release.